src/runtime.rs

```diff
@@ -153,7 +149,453 @@ pub struct FsSearchBody {
+    //     instance.loaded_directory_table_size += 1;
+    //     fs.folder_offset_count_1 += 1;
+    //     fs.file_info_index_count += new_info_indices.len() as u32;
+    //     fs.file_info_sub_index_count += new_info_to_datas.len() as u32;
+    //     fs.file_info_count += new_infos.len() as u32;
+    //     fs.file_data_count += new_datas.len() as u32;
+    //     instance.table2_len += new_info_indices.len() as u32;
+
+    //     Ok(())
+    // }
+
+    pub fn unshare_mass_loading_groups<Hash: Into<Hash40> + Clone>(paths: &Vec<Hash>) -> Result<(), LookupError> {
+        use std::collections::HashMap;
+        let paths: Vec<Hash40> = paths.iter().map(|x| {
+            x.clone().into()
+        }).collect();
+        unsafe {
+            // get the loaded structures
+            let mut instance = Self::acquire_instance(); // acquiring will lock the mutex and unlock on drop
+            let arc = Self::get_arc_mut();
+            let fs: &'static mut FileSystemHeader = std::mem::transmute(arc.fs_header);
+            let lengths = ArrayLengths::new(); // get array lengths as u32 values, simplifies making the indices
+
+            let folder_offsets = std::slice::from_raw_parts_mut(arc.folder_offsets as *mut DirectoryOffset, lengths.folder_offsets as usize);
+            let file_paths = std::slice::from_raw_parts_mut(arc.file_paths as *mut FilePath, lengths.file_paths as usize);
+            let file_info_indices = std::slice::from_raw_parts(arc.file_info_indices, lengths.file_info_indices as usize);
+            let file_infos = std::slice::from_raw_parts_mut(arc.file_infos, lengths.file_infos as usize);
+            let file_info_to_datas = std::slice::from_raw_parts(arc.file_info_to_datas, lengths.file_info_to_datas as usize);
+            let file_datas = std::slice::from_raw_parts(arc.file_datas, lengths.file_datas as usize);
+
+            let default_info = file_infos[0]; // default data to reserve sizes with
+
+            // declare the arrays to be filled in later
+            let mut new_info_indices = Vec::new();
+            let mut new_infos = Vec::new();
+            let mut new_info_to_datas = Vec::new();
+            let mut new_datas = Vec::new();
+            let mut new_mass_load_datas = Vec::new();
+
+            let mut mass_load_data_start_offset = 0;
+            for path in paths.iter() {
+                // start by changing the index of the load data
+                let mass_load_group = arc.get_dir_info_from_hash(*path)?;
+                let intermediate_load_data = &mut folder_offsets[(mass_load_group.dir_offset_index >> 8) as usize]; // ideally change this in smash-arc
+                let old_res_idx = intermediate_load_data.resource_index;
+                intermediate_load_data.resource_index = lengths.folder_offsets + new_mass_load_datas.len() as u32;
+                drop(intermediate_load_data); // can't mutably borrow twice at once, so drop
```

Comment:
shouldn't this drop not be needed? non-lexical lifetimes should automatically handle this case. does it compile without this drop?